File: packages/server/internal/model/device.go
package model

import "time"

// Device 设备表
// 记录信号灯监控设备的台账信息，hw_id 为出厂唯一硬件 ID
// 路口维度：intersection 为路口名称，lat/lng 为设备经纬度（用于地图打点）
type Device struct {
	ID           uint   `json:"id" gorm:"primaryKey"`
	HwID         string `json:"hw_id" gorm:"size:64;uniqueIndex;comment:设备硬件ID(uuid字符串,出厂唯一)"`
	Intersection string `json:"intersection" gorm:"size:128;index;comment:路口位置描述"`
	// P0-4 路口/行政区划挂接：均可空，只增不删，未挂接线旧设备不受影响
	CrossingID   *uint    `json:"crossing_id" gorm:"index;comment:所属路口ID(crossings.id,可空)"`
	ProvinceID   *uint    `json:"province_id" gorm:"comment:省ID"`
	CityID       *uint    `json:"city_id" gorm:"comment:市ID"`
	DistrictID   *uint    `json:"district_id" gorm:"comment:区县ID"`
	StreetID     *uint    `json:"street_id" gorm:"comment:街道ID"`
	CommunityID  *uint    `json:"community_id" gorm:"comment:社区ID"`
	RoadID       *uint    `json:"road_id" gorm:"comment:道路ID"`
	RoadName     string   `json:"road_name" gorm:"size:128;comment:道路名称(冗余)"`
	Lat          *float64 `json:"lat" gorm:"comment:纬度"`
	Lng          *float64 `json:"lng" gorm:"comment:经度"`
	NetworkCode  int      `json:"network_code" gorm:"comment:网络号"`
	StationCode  int      `json:"station_code" gorm:"comment:站点号"`
	SwVersion    uint32   `json:"sw_version" gorm:"comment:固件版本号"`
	ConfVersion  uint32   `json:"conf_version" gorm:"comment:配置版本号"`
	OnlineStatus bool     `json:"online_status" gorm:"default:false;comment:在线状态"`
	// 设备生命周期与接入状态：手工新增先预登记，首次 MQTT 上报后转为已接入。
	LifecycleStatus    string     `json:"lifecycle_status" gorm:"size:16;default:active;index;comment:生命周期(active/retired)"`
	AccessStatus       string     `json:"access_status" gorm:"size:16;default:never;index;comment:接入状态(never/accessed/offline)"`
	FirstAccessAt      *time.Time `json:"first_access_at" gorm:"comment:首次有效接入时间"`
	RetiredAt          *time.Time `json:"retired_at" gorm:"comment:报废时间"`
	RetiredReason      string     `json:"retired_reason" gorm:"size:255;comment:报废原因"`
	RegistrationSource string     `json:"registration_source" gorm:"size:16;default:manual;comment:建账来源(manual/mqtt_auto)"`
	IsWatched          bool       `json:"is_watched" gorm:"default:false;comment:是否关注(锁定/可能故障)"`
	LastCheckinAt      *time.Time `json:"last_checkin_at" gorm:"comment:最后签到时间"`
	InstalledAt        *time.Time `json:"installed_at" gorm:"comment:安装日期"`
	// 设备资料（照片/说明书/维修手册）：URL 可为外链或内部上传地址（/media/...），Name 为上传原文件名（用于阅读/下载展示）
	Photo            string `json:"photo" gorm:"size:500;comment:设备照片URL"`
	ManualUrl        string `json:"manual_url" gorm:"size:500;comment:说明书链接(外链或上传)"`
	ManualName       string `json:"manual_name" gorm:"size:255;comment:说明书文件名"`
	RepairManualUrl  string `json:"repair_manual_url" gorm:"size:500;comment:维修手册链接(外链或上传)"`
	RepairManualName string `json:"repair_manual_name" gorm:"size:255;comment:维修手册文件名"`
	// 对齐参考项目 a 的设备字段（a 的 equipment：func信号灯功能 / cx朝向 / fx方向 / batch批次 / remark备注）
	Func        string    `json:"func" gorm:"size:32;comment:信号灯功能(灯组类型,如直行/左转/右转)"`
	Orientation string    `json:"orientation" gorm:"size:16;comment:朝向(cx,如南/北/东/西/东南等)"`
	Direction   string    `json:"direction" gorm:"size:16;comment:方向(fx)"`
	Batch       string    `json:"batch" gorm:"size:32;comment:批次"`
	Remark      string    `json:"remark" gorm:"size:255;comment:备注"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// TableName 指定表名
func (Device) TableName() string {
	return "devices"
}

// 固件版本号位域解码（协议 PDF P6）
// bit[31:28]=major, bit[27:24]=minor, bit[23:18]=year(2000+n),
// bit[17:14]=month, bit[13:8]=day, bit[7:0]=build#
type SwVerInfo struct {
	Major uint32 `json:"major"`
	Minor uint32 `json:"minor"`
	Year  uint32 `json:"year"`
	Month uint32 `json:"month"`
	Day   uint32 `json:"day"`
	Build uint32 `json:"build"`
	Raw   uint32 `json:"raw"`
}

// DecodeSwVer 解码固件版本号
func DecodeSwVer(v uint32) SwVerInfo {
	return SwVerInfo{
		Major: (v >> 28) & 0xF,
		Minor: (v >> 24) & 0xF,
		Year:  (v >> 18) & 0x3F, // 6 bit，实际 2000+n
		Month: (v >> 14) & 0xF,
		Day:   (v >> 8) & 0x3F,
		Build: v & 0xFF,
		Raw:   v,
	}
}

// 配置版本号解码（协议 PDF P8）：0xYYMMDDnn
// YY=年 MM=月 DD=日 nn=当日版本
type ConfVerInfo struct {
	Year  uint32 `json:"year"`
	Month uint32 `json:"month"`
	Day   uint32 `json:"day"`
	Build uint32 `json:"build"`
	Raw   uint32 `json:"raw"`
}

// DecodeConfVer 解码配置版本号 0xYYMMDDnn
func DecodeConfVer(v uint32) ConfVerInfo {
	return ConfVerInfo{
		Year:  (v >> 24) & 0xFF,
		Month: (v >> 16) & 0xFF,
		Day:   (v >> 8) & 0xFF,
		Build: v & 0xFF,
		Raw:   v,
	}
}

File: packages/server-java/src/main/java/com/tsloms/server/model/Device.java
// 设备实体：逐列对齐 Go 版 internal/model/device.go（表 devices）。
package com.tsloms.server.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Index;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import java.time.Instant;

/** 信号灯监控设备台账；hw_id 为出厂唯一硬件 ID。 */
@Entity
@Table(name = "devices",
        uniqueConstraints = @UniqueConstraint(name = "uk_devices_hw_id", columnNames = "hw_id"),
        indexes = {
                @Index(name = "idx_devices_intersection", columnList = "intersection"),
                @Index(name = "idx_devices_crossing", columnList = "crossing_id"),
                @Index(name = "idx_devices_lifecycle", columnList = "lifecycle_status"),
                @Index(name = "idx_devices_access", columnList = "access_status"),
        })
public class Device extends BaseEntity {

    @Column(name = "hw_id", nullable = false, length = 64)
    public String hwId;

    /** 路口位置描述。 */
    @Column(name = "intersection", length = 128)
    public String intersection;

    // ---- P0-4 路口/行政区划挂接（均可空）----
    @Column(name = "crossing_id")
    public Long crossingId;

    @Column(name = "province_id")
    public Long provinceId;

    @Column(name = "city_id")
    public Long cityId;

    @Column(name = "district_id")
    public Long districtId;

    @Column(name = "street_id")
    public Long streetId;

    @Column(name = "community_id")
    public Long communityId;

    @Column(name = "road_id")
    public Long roadId;

    @Column(name = "road_name", length = 128)
    public String roadName;

    @Column(name = "lat")
    public Double lat;

    @Column(name = "lng")
    public Double lng;

    /** 网络号。 */
    @Column(name = "network_code", nullable = false)
    public Integer networkCode = 0;

    /** 站点号。 */
    @Column(name = "station_code", nullable = false)
    public Integer stationCode = 0;

    /** 固件版本号位域（uint32 → Long）。 */
    @Column(name = "sw_version", nullable = false)
    public Long swVersion = 0L;

    /** 配置版本号位域（uint32 → Long）。 */
    @Column(name = "conf_version", nullable = false)
    public Long confVersion = 0L;

    @Column(name = "online_status", nullable = false)
    public boolean onlineStatus;

    /** 生命周期：active/retired。 */
    @Column(name = "lifecycle_status", nullable = false, length = 16)
    public String lifecycleStatus = "active";

    /** 接入状态：never/accessed/offline。 */
    @Column(name = "access_status", nullable = false, length = 16)
    public String accessStatus = "never";

    @Column(name = "first_access_at")
    public Instant firstAccessAt;

    @Column(name = "retired_at")
    public Instant retiredAt;

    @Column(name = "retired_reason", length = 255)
    public String retiredReason;

    /** 建账来源：manual/mqtt_auto。 */
    @Column(name = "registration_source", nullable = false, length = 16)
    public String registrationSource = "manual";

    @Column(name = "is_watched", nullable = false)
    public boolean isWatched;

    @Column(name = "last_checkin_at")
    public Instant lastCheckinAt;

    @Column(name = "installed_at")
    public Instant installedAt;

    // ---- 设备资料 ----
    @Column(name = "photo", length = 500)
    public String photo;

    @Column(name = "manual_url", length = 500)
    public String manualUrl;

    @Column(name = "manual_name", length = 255)
    public String manualName;

    @Column(name = "repair_manual_url", length = 500)
    public String repairManualUrl;

    @Column(name = "repair_manual_name", length = 255)
    public String repairManualName;

    // ---- 参考项目 a 对齐字段 ----
    /** 信号灯功能（灯组类型：直行/左转/右转）。 */
    @Column(name = "func", length = 32)
    public String func;

    /** 朝向（南/北/东/西等）。 */
    @Column(name = "orientation", length = 16)
    public String orientation;

    @Column(name = "direction", length = 16)
    public String direction;

    @Column(name = "batch", length = 32)
    public String batch;

    @Column(name = "remark", length = 255)
    public String remark;
}
